typed/validate.go

```diff
@@ -0,0 +1,267 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/kubernetes-sigs/structured-merge-diff/fieldpath"
+	"github.com/kubernetes-sigs/structured-merge-diff/schema"
+	"github.com/kubernetes-sigs/structured-merge-diff/value"
+)
+
+type ValidationError struct {
+	Path  fieldpath.Path
+	Error string
+}
+
+func (ve ValidationError) FormatMessage() string {
+	return fmt.Sprintf("%s: %v", ve.Path, ve.Error)
+}
+
+// Validate returns an error with a list of every spec violation.
+func (tv TypedValue) Validate() error {
+	v := validation{
+		path:    fieldpath.Path{},
+		value:   tv.value,
+		schema:  tv.schema,
+		typeRef: tv.typeRef,
+	}
+	list := v.validate()
+	if len(list) == 0 {
+		return nil
+	}
+	messages := []string{"errors found:"}
+	for _, e := range list {
+		messages = append(messages, e.FormatMessage())
+	}
+	return errors.New(strings.Join(messages, "\n"))
+}
+
+type validation struct {
+	path    fieldpath.Path
+	value   value.Value
+	schema  *schema.Schema
+	typeRef schema.TypeRef
+}
+
+func (v validation) error(format string, args ...interface{}) ValidationError {
+	return ValidationError{
+		Path:  append(fieldpath.Path{}, v.path...),
+		Error: fmt.Sprintf(format, args...),
+	}
+}
+
+func (v validation) validate() []ValidationError {
+	a, ok := v.schema.Resolve(v.typeRef)
+	if !ok {
+		return []ValidationError{v.error("no type found matching: %v", *v.typeRef.NamedType)}
+	}
+
+	switch {
+	case a.Scalar != nil:
+		return v.doScalar(*a.Scalar, v.value)
+	case a.Struct != nil:
+		return v.doStruct(*a.Struct, v.value)
+	case a.List != nil:
+		return v.doList(*a.List, v.value)
+	case a.Map != nil:
+		return v.doMap(*a.Map, v.value)
+	case a.Untyped != nil:
+		// Untyped sections allow anything.
+		return nil
+	}
+
+	return []ValidationError{v.error("invalid atom")}
+}
+
+func (v validation) doScalar(t schema.Scalar, value value.Value) []ValidationError {
+	switch t {
+	case schema.Numeric:
+		if value.Float == nil && value.Int == nil {
+			// TODO: should the schema separate int and float?
+			return []ValidationError{v.error("expected numeric (int or float)")}
+		}
+	case schema.String:
+		if value.String == nil {
+			return []ValidationError{v.error("expected string")}
+		}
+	case schema.Boolean:
+		if value.Boolean == nil {
+			return []ValidationError{v.error("expected boolean")}
+		}
+	}
+	return nil
+}
+
+func (v validation) doStruct(t schema.Struct, value value.Value) (errs []ValidationError) {
+	switch {
+	case value.Null:
+		// Null is a valid struct.
+		return nil
+	case value.Map != nil:
+		// OK
+	default:
+		return []ValidationError{v.error("expected struct")}
+	}
+
+	allowedNames := map[string]struct{}{}
+	m := *value.Map
+	for i := range t.Fields {
+		f := &t.Fields[i]
```

Comment:
Do you need the pointer? Could you `range _, f := range t.Fields {` instead?

Reply:
I don't need the pointer but I don't want to use the loop variable since a reference might outlive the loop iteration (in an error message).